chips/nrf51/src/radio.rs

```diff
@@ -0,0 +1,456 @@
+                            self.set_data_white_iv(38);
+                            regs.READY.set(0);
+                            regs.TXEN.set(1);
+                        }
+                        // don't care as we only support advertisements at the moment
+                        _ => {
+                            self.set_channel_freq(37);
+                            self.set_data_white_iv(37)
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        self.enable_nvic();
+        self.enable_interrupts();
+    }
+
+
+    pub fn enable_interrupts(&self) {
+        // INTENSET REG
+        let regs = unsafe { &*self.regs };
+        // 15 i.e set 4 LSB
+        regs.INTENSET.set(1 | 1 << 1 | 1 << 2 | 1 << 3);
+    }
+
+    pub fn disable_interrupts(&self) {
+        let regs = unsafe { &*self.regs };
+        // disable all possible interrupts
+        regs.INTENCLR.set(0x4ff);
+    }
+
+    pub fn enable_nvic(&self) {
+        nvic::enable(NvicIdx::RADIO);
+    }
+
+    pub fn disable_nvic(&self) {
+        nvic::disable(NvicIdx::RADIO);
+    }
+
+    pub fn reset_payload(&self) {
+        // reset contents except header || address
+        for i in 9..39 {
+            unsafe {
+                PAYLOAD[i] = 0;
+            }
+        }
+        unsafe {
+            PAYLOAD[1] = 6;
+        }
+    }
+
+    // these are used by ble_advertising_driver and are therefore public
+    // FIXME: should be moved to the HIL later
+
+    pub fn start_adv(&self) {
+        self.init_radio_ble();
+    }
+
+    pub fn set_adv_data(&self,
+                        ad_type: usize,
+                        data: &'static mut [u8],
+                        len: usize,
+                        offset: usize)
+                        -> &'static mut [u8] {
+        if offset == 9 {
+            //FIXME: move call to the capsule!?
+            self.reset_payload();
+        }
+        unsafe {
+            PAYLOAD[offset] = (len + 1) as u8;
+            PAYLOAD[offset + 1] = ad_type as u8;
+        }
+
+        for (i, c) in data.as_ref()[0..len].iter().enumerate() {
+            unsafe {
+                PAYLOAD[i + offset + 2] = *c;
+            }
+        }
+        unsafe {
+            PAYLOAD[1] = (offset - 1 + len) as u8;
+        }
+        data
+    }
+
+    pub fn clear_adv_data(&self) {
+        self.reset_payload();
+    }
+
+    // FIXME: added a temporary variable in struct that keeps
+    // track of the twpower because we turn off the radio
+    // between advertisements,
+    // it is configured to 0 by default or the latest configured value
+    pub fn set_adv_txpower(&self, dbm: usize) -> ReturnCode {
+        match dbm {
+            // +4 dBm, 0 dBm, -4 dBm, -8 dBm, -12 dBm, -16 dBm, -20 dBm, -30 dBm
+            e @ 0x04 | e @ 0x00 | e @ 0xF4 | e @ 0xFC | e @ 0xF8 | e @ 0xF0 | e @ 0xEC |
+            e @ 0xD8 => {
+                self.txpower.set(e);
```

Comment:
Why not rewrite this as:

```
match dbm {
    // +4 dBm, 0 dBm, -4 dBm, -8 dBm, -12 dBm, -16 dBm, -20 dBm, -30 dBm
    0x04 | 0x00 | 0xF4 | 0xFC | 0xF8 | 0xF0 | 0xEC | 0xD8 => {
        self.txpower.set(dbm);
        ...
```

Does the `e @ 0x...` syntax actually serve any purpose?